mark.c

```diff
@@ -1651,15 +1683,45 @@ GC_API void GC_CALL GC_push_all_eager(void *bottom, void *top)
     if (top == 0) return;
 
     /* Check all pointers in range and push if they appear to be valid. */
-      lim = t - 1 /* longword */;
-      for (p = b; (word)p <= (word)lim;
-           p = (word *)(((ptr_t)p) + ALIGNMENT)) {
-        REGISTER word q = *p;
+    lim = t - 1 /* longword */;
+    for (p = b; (word)p <= (word)lim;
+         p = (word *)(((ptr_t)p) + ALIGNMENT)) {
+      REGISTER word q = *p;
 
-        GC_PUSH_ONE_STACK(q, p);
-      }
+      GC_PUSH_ONE_STACK(q, p);
+    }
+
+#   undef GC_greatest_plausible_heap_addr
+#   undef GC_least_plausible_heap_addr
+# else /* defined(__CHERI_PURE_CAPABILITY__) */
+
+    void **b = (void **)cheri_setaddress(bottom, (((vaddr_t) bottom + ALIGNMENT-1) 
+                                                 & ~(ALIGNMENT-1)));
+    void **t = (void **)cheri_setaddress(top, (((vaddr_t) top) 
+                                                 & ~(ALIGNMENT-1)));
```

Comment:
Suggestion:
```c++
void **b = (void**)cheri_align_up(bottom, ALIGNMENT);
void **t = (void**)cheri_align_down(bottom, ALIGNMENT);
```

Reply:
This is a good suggestion @jacobbramley .  Currently the code as it stands uses `cheri/cheric.h`  macros. 
This is I believe outdated.  
I have on my to-do list, converting all `cheri/cheric.h` macros to `cheriintrin,h`  macros.  I did not want to do it in this  PR and confuse matters. @ltratt , I would like to do this in a different PR right after this one. Would that be OK ? Or do you want it done here ?

Reply:
You can include both in the same file, I think, so you don't have to do it all at once.

I don't mind if you want to to it later, though.

Reply:
Interestingly, from my preliminary survey of CHERI memory allocators, both jemalloc and BSD's libmalloc_simple include `cheric.h`. I couldn't find any use of `cheriintrin.h` in allocators 'in the wild'.